agagd/agagd_core/views/players_profile.py

```diff
@@ -45,65 +42,79 @@ def get(self, request, *args, **kwargs):
-                tourney_data[game.tournament_code.pk] = t_dat
-            except ObjectDoesNotExist:
-                print("failing game_id: %s" % game.pk)
-
-        opp_table = PlayersOpponentTable(opponent_data.values())
-        RequestConfig(request, paginate={"per_page": 10}).configure(opp_table)
-
-        t_table = PlayersTournamentTable(
-            tourney_data.values(),
-            sorted(
-                tourney_data.values(),
-                key=lambda d: d.get("date", date.today()) or date.today(),
-                reverse=True,
-            ),
-            prefix="ts_played",
+        # Q objects to select games played by the player ...
+        Q_played = Q(pin_player_1__exact=player_id) | Q(pin_player_2__exact=player_id)
+        # ... won by the player ...
+        Q_won = Q(pin_player_1__exact=player_id, result__exact="W") | Q(
+            pin_player_2__exact=player_id, result__exact="B"
+        )
+        # ... and lost by the player
+        Q_lost = Q(pin_player_1__exact=player_id, result__exact="B") | Q(
+            pin_player_2__exact=player_id, result__exact="W"
+        )
+
+        opponents_queryset = (
+            agagd_models.Game.objects.filter(Q_played)
+            .annotate(
+                opponent_id=Case(
+                    When(
+                        pin_player_1__exact=player_id,
+                        then=F("pin_player_2"),
+                    ),
+                    When(
+                        pin_player_2__exact=player_id,
+                        then=F("pin_player_1"),
+                    ),
+                    output_field=IntegerField(),
+                ),
+                opponent_full_name=Case(
+                    When(
+                        pin_player_1__exact=player_id,
+                        then=F("pin_player_2__full_name"),
+                    ),
+                    When(
+                        pin_player_2__exact=player_id,
+                        then=F("pin_player_1__full_name"),
+                    ),
+                    output_field=CharField(),
+                ),
+            )
+            .values("opponent_id", "opponent_full_name")
+            .annotate(
+                won=Count("game_id", filter=Q_won),
+                lost=Count("game_id", filter=Q_lost),
+                total=F("won") + F("lost"),
+            )
+            .order_by("-total", "-won")
+        )
+        opponents_table = PlayersOpponentTable(opponents_queryset)
+        RequestConfig(request, paginate={"per_page": 10}).configure(opponents_table)
+
+        tournaments_queryset = (
+            agagd_models.Game.objects.filter(Q_played)
+            .values("tournament_code")
+            .annotate(
+                tournament_date=F("tournament_code__tournament_date"),
+                tournament_total_players=F("tournament_code__total_players"),
+                date=F("game_date"),
+                won=Count("game_id", filter=Q_won),
```

Comment:
This doesn't seem to be working -- all tournaments report just a single game.  Possibly missing a group-by?

Reply:
I don't see them always as being a single game with the fixture data, but I do see something fishy going on;  a bit hard to diagnose because the fixture data is not realistic, e.g., many players share the same name, and a tournament is made up from games from different dates.  Currently looking into this to see what is the problem.

Reply:
found the issue;  the entire aggregation was wrong because I was trying to construct the data starting from the Games table rather than the Tournament table.. Already have a fix;  I'll push it after I address the other feedback.

Reply:
Actually nvm;  I've pushed this one for now, since I'm not sure what to make of the other issue.  Latest commit should fix the tournament table in the player profile.